lib/signalfx/signalflow/computation.rb

```diff
@@ -0,0 +1,240 @@
+# handle.
+class Computation
+  attr_accessor :handle
+  attr_accessor :channel
+  attr_accessor :state
+  attr_accessor :metadata
+  attr_accessor :resolution
+  attr_accessor :input_timeseries_count
+
+  def initialize(handle, attach_func, stop_func)
+    @handle = handle
+    @channel = nil
+    @attach_func = attach_func
+    @stop_func = stop_func
+    @metadata = {}
+    # We can't have a handle until the job is started so we must be at least
+    # at this state
+    @state = STARTED_STATE
+
+    @pending_messages = Queue.new
+
+    @batch_size_known = false
+    @expected_batch_size = 0
+    @current_batch_data = nil
+    @current_batch_size = nil
+
+    @resolution = nil
+    @input_timeseries_count = nil
+  end
+
+  def channel=(channel)
+    @channel = channel
+  end
+
+  def attached?
+    !@channel.nil? && !@channel.detached
+  end
+
+  # Get the next message in this computation.
+  #
+  # @param timeout_seconds [Float] If a new message does not come within this
+  #   interval, raises a {ChannelTimeout} exception.  Note that this does not
+  #   mean that this function will return within this interval since there may
+  #   be messages received that are part of a larger batch.  If nil, will block
+  #   indefinitely.
+  def next_message(timeout_seconds=nil)
+    raise "Computation #{@handle} is not attached to a channel" unless @channel
+
+    msg = nil
+    while msg.nil? && !@channel.nil?
+      # process_message might return no messages if it is building up a batch
+      msg = process_message(@channel.pop(timeout_seconds))
+    end
+    return msg
+  end
+
+  # Iterates over the messages asynchronously for this computation.  A convenience
+  # function if you want to fire off multiple computations simultaneously,
+  # though not terribly efficient since it starts a new thread that spends a
+  # lot of time waiting.  However, since we don't have a way of "select"ing on
+  # computations, this is probably good enough for basic use.
+  #
+  # See {#each_message}.
+  def each_message_async(&block)
+    raise "Computation #{@handle} is not attached to a channel" unless @channel
+
+    Thread.new{ each_message(&block) }
+    return
+  end
+
+  # Call the given block with each message in the channel as they arrive.  This
+  # method will not return until the channel is detached from (either manually
+  # or due to the computation ending).
+  #
+  # Messages are queued in the channel so that none will be lost if this method
+  # is not called immediately.
+  #
+  # @yield [msg, comp] Called when a message arrives that is relevant to the
+  #   channel's computation.  The `comp` param will be set to this computation
+  #   instance for easy referencing of computation metadata and state.  `comp`
+  #   may be omitted if this reference to the computation is not needed.
+  def each_message(&block)
+    raise "Computation #{@handle} is not attached to a channel" unless @channel
+
+    while @channel
```

Comment:
I think you're missing the logic that transparently re-executes the computation if the channel stream ends before you reach the `COMPLETED_STATE`.